File: tactode-nodejs/tactode_pipeline.js
const cv = require('opencv4nodejs')
const params = require('./params.js')

// FUNCTIONS
function showContours (cnts = [], mat = new cv.Mat(), fld = '', apdx = '') {
  cnts = cnts.map(c => c.getPoints())
  for (let i = 0; i < cnts.length; ++i) {
    const color = new cv.Vec3(Math.random() * 255, Math.random() * 255, Math.random() * 255)
    mat.drawContours(cnts, i, color, 2, cv.LINE_AA)
  }
}

function drawRects (arr = [], mat = new cv.Mat(), fld = '', apdx = '') {
  arr.forEach(elem => {
    if (elem.length >= 1) {
      elem.forEach(subElem => {
        const color = new cv.Vec3(Math.random() * 255, Math.random() * 255, Math.random() * 255)
        mat.drawRectangle(subElem.rect, color, 2)
      })
    } else {
      const color = new cv.Vec3(Math.random() * 255, Math.random() * 255, Math.random() * 255)
      mat.drawRectangle(elem.rect, color, 2)
    }
  })
}

function unionRect (rect1 = new cv.Rect(), rect2 = new cv.Rect()) {
  const top1 = rect1.y; const bottom1 = rect1.y + rect1.height
  const left1 = rect1.x; const right1 = rect1.x + rect1.width
  const top2 = rect2.y; const bottom2 = rect2.y + rect2.height
  const left2 = rect2.x; const right2 = rect2.x + rect2.width

  const top = (top1 <= top2) ? top1 : top2
  const bottom = (bottom1 >= bottom2) ? bottom1 : bottom2
  const left = (left1 <= left2) ? left1 : left2
  const right = (right1 >= right2) ? right1 : right2

  const width = right - left
  const height = bottom - top

  return new cv.Rect(left, top, width, height)
}

function toSquare (rect = new cv.Rect()) {
  const diff = Math.round(Math.abs(rect.width - rect.height) / 2)
  const side = (rect.width >= rect.height) ? rect.width : rect.height
  const x = (rect.width >= rect.height) ? rect.x : rect.x - diff
  const y = (rect.height >= rect.width) ? rect.y : rect.y - diff

  return new cv.Rect(x, y, side, side)
}

const newContour = (idx = -1, hierarchy = new cv.Vec4(-1, -1, -1, -1), rect = new cv.Rect(),
  center = new cv.Point2(-1, -1), area = -1, perimeter = -1,
  aspectRatio = -1, extent = -1, solidity = -1, equiDiameter = -1) => {
  return {
    idx: idx,
    haveTeeth: false,
    hierarchy: hierarchy,
    rect: rect,
    center: {
      x: center.x,
      y: center.y
    },
    properties: {
      area: area,
      perimeter: perimeter,
      aspectRatio: aspectRatio,
      extent: extent,
      solidity: solidity,
      equiDiameter: equiDiameter
    },
    updateCenter: function () {
      this.center.x = this.rect.x + Math.round(this.rect.width / 2)
      this.center.y = this.rect.y + Math.round(this.rect.height / 2)
    }
  }
}

const newPiece = (rect = new cv.Rect()) => {
  return {
    rect: rect,
    tactode: {
      type: '',
      piece: ''
    }
  }
}

function runMain (imgPath = '', hog = new cv.HOGDescriptor(), svm = new cv.SVM(), config = new params.HogSvmDetector()) {
  // -> Beginning
  let src = cv.imread(imgPath)

  // -> Resizing image
  const normSize = 1920
  let rows = 0; let cols = 0; let aspectRatio = 0
  if (src.rows !== normSize && src.cols !== normSize) {
    if (src.rows >= src.cols) {
      aspectRatio = src.rows / src.cols
      rows = normSize
      cols = Math.floor(normSize / aspectRatio)
    } else {
      aspectRatio = src.cols / src.rows
      cols = normSize
      rows = Math.floor(normSize / aspectRatio)
    }
    src = src.resize(rows, cols, 0, 0, cv.INTER_AREA)
  }

  // -> Adjusting image
  let adjusted = src.copy(src)
  adjusted = cv.gaussianBlur(adjusted, new cv.Size(9, 9), 0, 0, cv.BORDER_DEFAULT)
  adjusted = adjusted.resize(0, 0, 0.3, 0.3, cv.INTER_AREA)
  adjusted = adjusted.cvtColor(cv.COLOR_BGR2GRAY, 0).canny(100, 200, 3)
  const lines = adjusted.houghLines(1, Math.PI / 180, 50, 0, 0, 0, Math.PI)
  if (lines.length > 0) {
    let angle = -1
    const bigAngles = []
    for (let i = 0; i < lines.length; ++i) {
      const degrees = lines[i].y * 180 / Math.PI
      if (angle === -1) {
        angle = degrees
        bigAngles.push(degrees)
      } else {
        if (Math.abs(degrees - angle) > 40 && Math.abs(degrees - angle) < 140) {
          bigAngles.push(degrees)
          break
        }
      }
    }
    bigAngles.forEach(a => {
      let diffAngle = a
      const idx = bigAngles.indexOf(a)
      if (diffAngle > 45 && diffAngle <= 135) {
        diffAngle = diffAngle - 90
      } else if (diffAngle > 135) {
        diffAngle = diffAngle - 180
      } else {
        diffAngle += 0
      }
      bigAngles.splice(idx, 1, diffAngle)
    })
    let adjustingAngle = -1
    if (bigAngles.length > 1) {
      if (bigAngles[0] * bigAngles[1] < 0) {
        adjustingAngle = (bigAngles[0] - bigAngles[1]) / 2
      } else {
        adjustingAngle = (bigAngles[0] + bigAngles[1]) / 2
      }
    } else {
      adjustingAngle = bigAngles[0]
    }
    const rotCenter = new cv.Point2(src.cols / 2, src.rows / 2)
    const rotMatrix = cv.getRotationMatrix2D(rotCenter, adjustingAngle, 1)
    const dsize = new cv.Size(src.cols, src.rows)
    src = src.warpAffine(rotMatrix, dsize, cv.INTER_LINEAR, cv.BORDER_DEFAULT, new cv.Vec3(0, 0, 0))
  }

  // -> Binarising image
  const blur = cv.gaussianBlur(src, new cv.Size(11, 11), 0, 0, cv.BORDER_DEFAULT)
  const kernel = cv.getStructuringElement(cv.MORPH_CROSS, new cv.Size(6, 6))
  const grad = blur.morphologyEx(kernel, cv.MORPH_GRADIENT)
  const bw = grad.inRange(new cv.Vec3(0, 0, 0), new cv.Vec3(20, 20, 20))
  let fg = bw.morphologyEx(kernel, cv.MORPH_OPEN)
  fg = fg.morphologyEx(kernel, cv.MORPH_CLOSE)

  // -> Segmenting and rotating image
  let rotationStep = 0
  let isRotated = true
  let bigger = newContour()
  let smaller = newContour()
  let cntArr = []
  let contours = [new cv.Contour()]
  while (rotationStep < 4 && isRotated) {
    cntArr = []
    bigger = newContour()
    smaller = newContour()

    // -> Finding contours
    contours = fg.findContours(cv.RETR_TREE, cv.CHAIN_APPROX_NONE, new cv.Point2(0, 0))

    // -> Getting contours' properties
    for (let i = 0; i < contours.length; ++i) {
      const cnt = contours[i]
      const hier = cnt.hierarchy

      const rect = cnt.boundingRect()
      const hull = cnt.convexHull(false)

      const rectArea = rect.width * rect.height
      const center = new cv.Point2(rect.x + rect.width / 2, rect.y + rect.height / 2)

      const area = cnt.area
      const perimeter = cnt.arcLength(false)
      const aspectRatio = rect.width / rect.height
      const extent = area / rectArea
      const solidity = area / hull.area
      const equiDiameter = Math.sqrt(4 * area / Math.PI)

      cntArr.push(newContour(i, hier, rect, center, area, perimeter,
        aspectRatio, extent, solidity, equiDiameter))
    }

    // -> Finding grandparent, parent and child contours
    cntArr.sort((c0, c1) => c1.properties.area - c0.properties.area)
    let grandparentidx = -1
    let parentidx = -1
    for (let i = 0; i < cntArr.length; ++i) {
      const ch = cntArr[i].hierarchy
      const ci = cntArr[i].idx
      if (grandparentidx === -1) {
        if (ch.z === -1 && ch.y !== -1) {
          grandparentidx = ci
        }
      } else {
        if (ch.z === grandparentidx) {
          parentidx = ci
          break
        }
      }
    }
    cntArr = cntArr.filter(c => c.hierarchy.z === parentidx)

    // -> Finding biggest and smallest tile-alike contours
    for (let i = 0; i < cntArr.length; ++i) {
      const c = cntArr[i]
      if (bigger.idx === -1 && c.properties.aspectRatio >= 0.7) {
        bigger = c
      } else if (bigger.idx !== -1 && c.properties.area / bigger.properties.area < 0.35) {
        smaller = cntArr[i - 1]
        break
      }
    }

    // -> Detecting teeth in the tiles
    for (let i = cntArr.indexOf(bigger); i <= cntArr.indexOf(smaller); ++i) {
      const c = cntArr[i]
      const width = c.rect.width
      let height = Math.round(c.rect.height / 3)
      const x = c.rect.x
      const y = c.rect.y + 2 * height
      height = Math.round(width / 2)
      const roi = src.getRegion(new cv.Rect(x, y, width, height))
        .resize(config.size.height, config.size.width, 0, 0, cv.INTER_AREA)
      const descriptors = hog.compute(roi)
      const prediction = svm.predict(descriptors)
      c.haveTeeth = (prediction === 1)
    }

    // -> Filtering teeth detections and identifying rotation
    const teethArr = cntArr.filter(c => c.haveTeeth)
    if (teethArr.length > 0) {
      teethArr.sort((t0, t1) => t0.rect.x - t1.rect.x)
      const mostLeft = teethArr[0]
      if (!cntArr.some(c => c.rect.x < mostLeft.rect.x - mostLeft.rect.width / 2)) {
        let prevTeeth = newContour()
        let columnCount = 1
        for (let i = 0; i < teethArr.length; ++i) {
          const currTeeth = teethArr[i]
          if (prevTeeth.idx === -1) {
            prevTeeth = currTeeth
          } else {
            if (Math.abs(prevTeeth.rect.x - currTeeth.rect.x) <= prevTeeth.rect.width / 2) {
              columnCount++
              prevTeeth = currTeeth
            }
            if (columnCount >= 2) {
              isRotated = false
              break
            }
          }
        }
      }
    }
    if (isRotated) {
      fg = fg.rotate(cv.ROTATE_90_CLOCKWISE)
      src = src.rotate(cv.ROTATE_90_CLOCKWISE)
    }
    rotationStep++
  }

  // -> Removing inner and noisy contours
  for (let i = 0; i < cntArr.length; ++i) {
    const currCnt = cntArr[i]
    if (currCnt.idx !== -1) {
      const currRectArea = currCnt.rect.width * currCnt.rect.height
      const subArr = cntArr.slice(i + 1)
      for (let j = 0; j < subArr.length; ++j) {
        const subC = subArr[j]
        if (subC.idx !== -1) {
          const resOrRect = unionRect(currCnt.rect, subC.rect)
          const resOrRectArea = resOrRect.width * resOrRect.height
          const idx = i + j + 1
          if (resOrRectArea / currRectArea === 1 || subC.properties.aspectRatio < 0.25 || subC.properties.aspectRatio > 5 || subC.properties.area / bigger.properties.area < 0.01 || subC.properties.extent <= 0.25) {
            contours.splice(cntArr[idx].idx, 1, new cv.Contour())
            cntArr.splice(idx, 1, newContour())
          }
        }
      }
    }
  }
  cntArr = cntArr.filter(c => c.idx !== -1)

  // -> Agglomerating little bits of remaining contours
  cntArr.sort((c0, c1) => c0.properties.area - c1.properties.area)

  for (let i = 0; i < cntArr.length; ++i) {
    const currCnt = cntArr[i]
    const currRectArea = currCnt.rect.width * currCnt.rect.height
    const currCenter = currCnt.center
    const slicedArr = cntArr.slice(i + 1)
    let subsubArr = [newContour()]
    let idx = -1
    for (let j = 0; j < slicedArr.length; ++j) {
      const subCnt = slicedArr[j]
      const subRectArea = subCnt.rect.width * subCnt.rect.height
      if (currRectArea / subRectArea < 0.25) {
        subsubArr = slicedArr.slice(j)
        idx = j
        break
      }
    }
    if (idx !== -1) {
      contours.splice(cntArr[i].idx, 1, new cv.Contour())
      cntArr.splice(i, 1, newContour())
      subsubArr.sort((c0, c1) => {
        const c0x = Math.pow((c0.center.x - currCenter.x), 2)
        const c0y = Math.pow((c0.center.y - currCenter.y), 2)
        const c1x = Math.pow((c1.center.x - currCenter.x), 2)
        const c1y = Math.pow((c1.center.y - currCenter.y), 2)
        const leftSide = Math.sqrt(c0x + c0y)
        const rightSide = Math.sqrt(c1x + c1y)
        const dist = Math.round(leftSide - rightSide)
        return dist
      })
      let closerElem = newContour()
      let next = false
      for (let k = 0; k < subsubArr.length; ++k) {
        const leftLimit = subsubArr[k].rect.x
        const rightLimit = subsubArr[k].rect.x + subsubArr[k].rect.width
        const topLimit = subsubArr[k].rect.y
        const bottomLimit = subsubArr[k].rect.y + subsubArr[k].rect.height
        if (currCenter.x >= leftLimit && currCenter.x <= rightLimit) {
          if (currCenter.y < subsubArr[k].center.y || (currCenter.y <= bottomLimit && currCenter.y >= topLimit)) {
            closerElem = subsubArr[k]
            break
          } else if (next) {
            closerElem = subsubArr[k - 1]
            break
          } else {
            next = true
          }
        } else {
          if (currCenter.y <= bottomLimit && currCenter.y >= topLimit) {
            closerElem = subsubArr[k]
            break
          } else if (next) {
            closerElem = subsubArr[k - 1]
            break
          } else {
            next = true
          }
        }
      }
      const otheridx = cntArr.indexOf(closerElem)
      if (otheridx !== -1) {
        cntArr[otheridx].rect = unionRect(cntArr[otheridx].rect, currCnt.rect)
        cntArr[otheridx].updateCenter()
      }
    }
  }
  cntArr = cntArr.filter(c => c.idx !== -1)

  // -> Positioning
  cntArr.sort((c0, c1) => c0.rect.y - c1.rect.y)
  const rectsArr = []; let rLine = []
  let prevCnt
  cntArr.forEach(currCnt => {
    if (prevCnt === undefined) {
      prevCnt = currCnt
    } else {
      if (currCnt.rect.y > prevCnt.rect.y + prevCnt.rect.height / 4) { // new line found
        rLine.push(prevCnt)
        rLine.sort((c0, c1) => c0.rect.x - c1.rect.x)
        rLine = rLine.map(c => newPiece(toSquare(c.rect)))
        rectsArr.push(rLine)
        rLine = []
      } else {
        rLine.push(prevCnt)
      }
      if (cntArr.indexOf(currCnt) >= cntArr.length - 1) { // last piece
        rLine.push(currCnt)
        rLine.sort((c0, c1) => c0.rect.x - c1.rect.x)
        rLine = rLine.map(c => newPiece(toSquare(c.rect)))
        rectsArr.push(rLine)
      }
      prevCnt = currCnt
    }
  })

  // -> Returning result
  return new params.PipelineResult(rectsArr, src, isRotated)
}

module.exports = {
  runMain
}

File: tactode-python/tactode_pipeline.py
import cv2 as cv
import numpy as np
import math


# CLASSES
class Center:
    def __init__(self, xx=-1, yy=-1):
        self.x = xx
        self.y = yy


class Rect:
    def __init__(self, xx=-1, yy=-1, hc=-1, wc=-1):
        self.x = int(xx)
        self.y = int(yy)
        self.height = int(hc)
        self.width = int(wc)


class Contour:
    def __init__(self, ic=-1, hc=None, r=Rect(), cc=Center(),
                 ar=None, per=None, asp_rat=None, ext=None, equi_diam=None):
        self.idx = ic
        self.have_teeth = False
        self.hierarchy = hc
        self.rect = r
        self.center = cc
        self.area = ar
        self.perimeter = per
        self.aspect_ratio = asp_rat
        self.extent = ext
        self.equi_diameter = equi_diam

    def update_center(self):
        self.center.x = self.rect.x + round(self.rect.width / 2)
        self.center.y = self.rect.y + round(self.rect.height / 2)


# FUNCTIONS
def draw_contours(img, arr, contours):
    for e in range(len(arr)):
        img = cv.drawContours(img, contours, arr[e].idx, (0, 255, 0), 6)
    return img


def draw_rects(img, arr):
    for r_elem in arr:
        cv.rectangle(img, (r_elem.rect.x, r_elem.rect.y),
                     (r_elem.rect.x + r_elem.rect.width, r_elem.rect.y + r_elem.rect.height),
                     (0, 255, 255), 6)
    return img


def to_square(rect=Rect()):
    diff = round(abs(rect.width - rect.height) / 2)
    side = rect.width if rect.width >= rect.height else rect.height
    x_r = rect.x if rect.width >= rect.height else rect.x - diff
    y_r = rect.y if rect.height >= rect.width else rect.y - diff
    return Rect(x_r, y_r, side, side)


def union_rect(r1=Rect(), r2=Rect()):
    top1 = r1.y
    bottom1 = r1.y + r1.height
    left1 = r1.x
    right1 = r1.x + r1.width
    top2 = r2.y
    bottom2 = r2.y + r2.height
    left2 = r2.x
    right2 = r2.x + r2.width

    top = top1 if top1 <= top2 else top2
    bottom = bottom1 if bottom1 >= bottom2 else bottom2
    left = left1 if left1 <= left2 else left2
    right = right1 if right1 >= right2 else right2

    width = right - left
    height = bottom - top

    return Rect(left, top, height, width)


def run_main(img_path, hog, svm, config):
    # Reading input image
    src = cv.imread(img_path)
    old_height = src.shape[0]
    old_width = src.shape[1]

    # Resizing image
    norm_size = 1920
    new_height, new_width, aspect_ratio = 0, 0, 0
    if old_height != norm_size and old_width != norm_size:
        if old_height >= old_width:
            aspect_ratio = old_height / old_width
            new_height = norm_size
            new_width = int(np.floor(norm_size / aspect_ratio))
        else:
            aspect_ratio = old_width / old_height
            new_width = norm_size
            new_height = int(np.floor(norm_size / aspect_ratio))
        src = cv.resize(src, (new_width, new_height), cv.INTER_AREA)
    # cv.imshow("resized", cv.resize(src.copy(), dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()
    # cv.imwrite("../../use_case_images/img-1.jpg", src)

    # Adjusting image
    to_adjust = src.copy()
    to_adjust = cv.GaussianBlur(to_adjust, (9, 9), 0)
    to_adjust = cv.resize(to_adjust, None, fx=0.3, fy=0.3, interpolation=cv.INTER_AREA)
    to_adjust = cv.cvtColor(to_adjust, cv.COLOR_BGR2GRAY)
    to_adjust = cv.Canny(to_adjust, 100, 200, apertureSize=3)
    hough_lines = cv.HoughLines(to_adjust, rho=1, theta=np.pi / 180, threshold=50,
                                srn=0, stn=0, min_theta=0, max_theta=np.pi)

    if hough_lines.shape[0] > 0:
        angle = -1
        big_angles = []
        found_biggest_lines = False
        for row in range(0, len(hough_lines)):
            for rho, theta in hough_lines[row]:
                degrees = theta * 180 / np.pi
                if angle == -1:
                    angle = degrees
                    big_angles.append(degrees)
                else:
                    if 40 < np.abs(degrees - angle) < 140:
                        big_angles.append(degrees)
                        found_biggest_lines = True
            if found_biggest_lines:
                break

    for a in big_angles:
        diff_angle = a
        idx_big_angle = big_angles.index(a)
        if 45 < diff_angle <= 135:
            diff_angle -= 90
        elif diff_angle > 135:
            diff_angle -= 180
        else:
            diff_angle += 0
        big_angles.pop(idx_big_angle)
        big_angles.insert(idx_big_angle, diff_angle)

    if len(big_angles) > 1:
        if big_angles[0] * big_angles[1] < 0:
            adjusting_angle = (big_angles[0] - big_angles[1]) / 2
        else:
            adjusting_angle = (big_angles[0] + big_angles[1]) / 2
    else:
        adjusting_angle = big_angles[0]

    h = src.shape[0]
    w = src.shape[1]
    rangle = np.deg2rad(adjusting_angle)    # angle in radians
    # Compute new image height and width
    nh = (abs(np.cos(rangle) * h) + abs(np.sin(rangle) * w)) * 1.
    nw = (abs(np.sin(rangle) * h) + abs(np.cos(rangle) * w)) * 1.
    # Compute rotation matrix
    rot_mat = cv.getRotationMatrix2D((nw*.5, nh*.5), adjusting_angle, 1.)
    # Compute the move from the old center to the new center combined w/ no rotation
    rot_move = np.dot(rot_mat, np.array([(nw-w)*.5, (nh-h)*.5, 0]))
    # Update the translation
    rot_mat[0, 2] += rot_move[0]
    rot_mat[1, 2] += rot_move[1]
    src = cv.warpAffine(src, rot_mat, (int(math.ceil(nw)), int(math.ceil(nh))), flags=cv.INTER_LANCZOS4)
    # cv.imshow("adjusted", cv.resize(src.copy(), dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()
    # cv.imwrite("../../use_case_images/img-2.jpg", src)

    # Binarising image
    blur = cv.GaussianBlur(src, (11, 11), 0)
    kernel = cv.getStructuringElement(cv.MORPH_CROSS, (6, 6))
    grad = cv.morphologyEx(blur, cv.MORPH_GRADIENT, kernel)
    bw = cv.inRange(grad, (0, 0, 0), (20, 20, 20))
    fg = cv.morphologyEx(bw, cv.MORPH_OPEN, kernel)
    fg = cv.morphologyEx(fg, cv.MORPH_CLOSE, kernel)
    # cv.imshow("binary", cv.resize(fg, dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()
    # cv.imwrite("../../use_case_images/img-3.jpg", fg)

    # Segmenting and rotating image
    rotation_step = 0
    is_rotated = True
    cnt_arr = []
    child_cnt_arr = []
    teeth_arr = []
    biggest_cnt = Contour()

    while rotation_step < 4 and is_rotated:
        biggest_cnt = Contour()
        smallest_cnt = Contour()
        cnt_arr.clear()

        # Finding contours
        contours, hierarchy = cv.findContours(fg, cv.RETR_TREE, cv.CHAIN_APPROX_NONE)
        # immm, contours, hierarchy = cv.findContours(fg, cv.RETR_TREE, cv.CHAIN_APPROX_NONE)

        # Getting contours' properties
        for i in range(0, len(contours)):
            cnt = contours[i]
            hier = hierarchy[0][i]

            x, y, w, h = cv.boundingRect(cnt)

            rect_area = int(w * h)
            center = Center(int(x + w / 2), int(y + h / 2))

            area = int(cv.contourArea(cnt))
            perimeter = cv.arcLength(cnt, True)
            cnt_aspect_ratio = w / h
            extent = area / rect_area
            equi_diameter = np.sqrt(4 * area / np.pi)

            cnt_arr.append(Contour(i, hier, Rect(x, y, h, w), center,
                                   area, perimeter, cnt_aspect_ratio, extent, equi_diameter))

        # Finding grandparent, parent and child contours
        cnt_arr.sort(key=lambda c: c.area, reverse=True)
        grandparent_idx = -1
        parent_idx = -1
        for c in cnt_arr:
            ch = c.hierarchy
            ci = c.idx
            if grandparent_idx == -1:
                if ch[3] == -1 and ch[2] != -1:
                    grandparent_idx = ci
            else:
                if ch[3] == grandparent_idx:
                    parent_idx = ci
                    break

        child_cnt_arr.clear()
        for c in cnt_arr:
            if c.hierarchy[3] == parent_idx:
                child_cnt_arr.append(c)

        # Finding biggest and smallest tile-alike contours
        for i in range(0, len(child_cnt_arr)):
            c = child_cnt_arr[i]
            if biggest_cnt.idx == -1 and c.aspect_ratio >= 0.7:
                biggest_cnt = c
            elif biggest_cnt != -1 and c.area / biggest_cnt.area < 0.35:
                smallest_cnt = child_cnt_arr[i - 1]
                break

        # Detecting teeth in the tiles
        for i in range(child_cnt_arr.index(biggest_cnt), child_cnt_arr.index(smallest_cnt) + 1):
            cnt = child_cnt_arr[i]
            width = int(cnt.rect.width)
            height = round(cnt.rect.height / 3)
            x = int(cnt.rect.x)
            y = int(cnt.rect.y + 2 * height)
            height = int(round(width / 2))
            roi = src[y: y + height, x: x + width]
            roi = cv.resize(roi, (config.size.width, config.size.height), cv.INTER_AREA)
            sample = []
            histograms = hog.compute(roi)
            sample.append(histograms)
            sample = np.float32(sample)
            label = svm.predict(sample)
            label = int(label[1][0])
            cnt.have_teeth = True if label == 1 else False

        # Filtering teeth detections and identifying rotation
        teeth_arr.clear()
        for c in child_cnt_arr:
            if c.have_teeth:
                teeth_arr.append(c)
        if len(teeth_arr) > 0:
            teeth_arr.sort(key=lambda c: c.rect.x, reverse=False)
            most_left_tile = teeth_arr[0]
            tiles_at_left = False
            for c in child_cnt_arr:
                if c.rect.x < most_left_tile.rect.x - most_left_tile.rect.width / 2:
                    tiles_at_left = True
                    break
            if not tiles_at_left:
                prev_teeth = Contour()
                column_count = 1
                for curr_teeth in teeth_arr:
                    if prev_teeth.idx == -1:
                        prev_teeth = curr_teeth
                    else:
                        if np.abs(prev_teeth.rect.x - curr_teeth.rect.x) <= prev_teeth.rect.width / 2:
                            column_count += 1
                            prev_teeth = curr_teeth
                        if column_count >= 2:
                            is_rotated = False
                            break
        if is_rotated:
            fg = cv.rotate(fg, cv.ROTATE_90_CLOCKWISE)
            src = cv.rotate(src, cv.ROTATE_90_CLOCKWISE)
        rotation_step += 1
    # cv.imshow("rotated", cv.resize(draw_contours(src.copy(), child_cnt_arr, contours),
    #                               dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()
    # cv.imwrite("../../use_case_images/img-4.jpg", draw_contours(src.copy(), child_cnt_arr, contours))

    # Removing inner and noisy contours
    sub_arr = child_cnt_arr.copy()
    for i in range(0, len(child_cnt_arr)):
        curr_cnt = child_cnt_arr[i]
        if curr_cnt.idx != -1:
            curr_rect_area = int(curr_cnt.rect.width * curr_cnt.rect.height)
            for j in range(i + 1, len(sub_arr)):
                sub_cnt = sub_arr[j]
                if sub_cnt.idx != -1:
                    result_union_rect = union_rect(curr_cnt.rect, sub_cnt.rect)
                    result_union_rect_area = int(result_union_rect.width * result_union_rect.height)
                    if result_union_rect_area == curr_rect_area or \
                            sub_cnt.aspect_ratio < 0.25 or sub_cnt.aspect_ratio > 5 or \
                            sub_cnt.area / biggest_cnt.area < 0.01 or sub_cnt.extent <= 0.25:
                        child_cnt_arr[j].idx = -1

    aux_cnt_arr = []
    for c in child_cnt_arr:
        if c.idx != -1:
            aux_cnt_arr.append(c)
    child_cnt_arr.clear()
    child_cnt_arr = aux_cnt_arr.copy()
    aux_cnt_arr.clear()
    # cv.imshow("without noisy cnts", cv.resize(draw_contours(src.copy(), child_cnt_arr, contours),
    #                                          dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()

    # Agglomerating little bits of remaining contours
    def sort_by_center_distance(ccd=Contour()):
        cx = pow((ccd.center.x - curr_center.x), 2)
        cy = pow((ccd.center.y - curr_center.y), 2)
        dist = round(np.sqrt(cx + cy))
        return dist

    sub_arr.clear()
    child_cnt_arr.sort(key=lambda c: c.area, reverse=False)
    sub_arr = child_cnt_arr.copy()
    for i in range(0, len(child_cnt_arr)):
        curr_cnt = child_cnt_arr[i]
        curr_rect_area = curr_cnt.rect.width * curr_cnt.rect.height
        curr_center = curr_cnt.center
        sub_sub_arr = [Contour()]
        idx = -1
        for j in range(i + 1, len(sub_arr)):
            sub_cnt = sub_arr[j]
            sub_rect_area = sub_cnt.rect.width * sub_cnt.rect.height
            if curr_rect_area / sub_rect_area < 0.25:
                sub_sub_arr = sub_arr[j:]
                idx = j
                break
        if idx != -1:
            child_cnt_arr[i].idx = -1
            sub_sub_arr.sort(key=sort_by_center_distance, reverse=False)
            closest_cnt = Contour()
            next_cnt = False
            for k in range(0, len(sub_sub_arr)):
                left_limit = sub_sub_arr[k].rect.x
                right_limit = sub_sub_arr[k].rect.x + sub_sub_arr[k].rect.width
                top_limit = sub_sub_arr[k].rect.y
                bottom_limit = sub_sub_arr[k].rect.y + sub_sub_arr[k].rect.height
                if left_limit <= curr_center.x <= right_limit:
                    if curr_center.y < sub_sub_arr[k].center.y or (top_limit <= curr_center.y <= bottom_limit):
                        closest_cnt = sub_sub_arr[k]
                        break
                    elif next_cnt:
                        closest_cnt = sub_sub_arr[k - 1]
                        break
                    else:
                        next_cnt = True
                else:
                    if top_limit <= curr_center.y <= bottom_limit:
                        closest_cnt = sub_sub_arr[k]
                        break
                    elif next_cnt:
                        closest_cnt = sub_sub_arr[k - 1]
                        break
                    else:
                        next_cnt = True
            other_idx = child_cnt_arr.index(closest_cnt)
            if other_idx != -1:
                child_cnt_arr[other_idx].rect = union_rect(child_cnt_arr[other_idx].rect, curr_cnt.rect)
                child_cnt_arr[other_idx].update_center()

    for c in child_cnt_arr:
        if c.idx != -1:
            aux_cnt_arr.append(c)
    child_cnt_arr.clear()
    child_cnt_arr = aux_cnt_arr.copy()
    aux_cnt_arr.clear()

    # cv.imwrite("out.jpg", draw_rects(src.copy(), child_cnt_arr))
    # print("Number of tiles found: ", len(child_cnt_arr))
    # cv.imshow("agglomerated", cv.resize(draw_contours(src.copy(), child_cnt_arr, contours),
    #                             dsize=(0, 0), fx=0.3, fy=0.3))
    # cv.waitKey()
    # cv.imwrite("../../use_case_images/img-5.jpg", draw_contours(src.copy(), child_cnt_arr, contours))

    # Positioning
    child_cnt_arr.sort(key=lambda c: c.rect.y, reverse=False)
    tiles = []
    line = []
    prev_cnt = Contour()
    for curr_cnt in child_cnt_arr:
        if prev_cnt.idx == -1:
            prev_cnt = curr_cnt
        else:
            if curr_cnt.rect.y > prev_cnt.rect.y + prev_cnt.rect.height / 4:
                line.append(to_square(prev_cnt.rect))
                line.sort(key=lambda r: r.x, reverse=False)
                aux_line = line.copy()
                tiles.append(aux_line)
                line.clear()
            else:
                line.append(to_square(prev_cnt.rect))
            if child_cnt_arr.index(curr_cnt) >= len(child_cnt_arr) - 1:
                line.append(to_square(curr_cnt.rect))
                line.sort(key=lambda r: r.x, reverse=False)
                aux_line = line.copy()
                tiles.append(aux_line)
            prev_cnt = curr_cnt
    return tiles, src, is_rotated
